proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,622 @@
+```
+
+Definition for transaction header parameters:
+
+```
+enum TransactionHeaderParameter {
+  // existing compute budget parameters
+  RequestHeapFrame(u32),
+  SetComputeUnitLimit(u32),
+  SetComputeUnitPrice(u64),
+  SetLoadedAccountsDataSizeLimit(u32),
+  // Limit on maximum of PRAW fees to be paid
+  // Fees is set in microlamports per CU
+  SetProgramRebatableAccountWriteFees(u64),
+}
+```
+
+When the cluster receives a new transaction, with `TransactionHeaderParameter`
+instruction is decoded to get the `SetProgramRebatableAccountWriteFees` and then
+calculates the total fees required for the transaction message. Then we verify
+that the fee-payer has a minimum balance of: `per-transaction base fees` +
+`prioritization fees` +
+`Microlamports per CU set by SetProgramRebatableAccountWriteFees` *
+`requested CUs set by SetComputeUnitLimit`.
+
+Where PRAW fees is :
+`Microlamports per CU set by SetProgramRebatableAccountWriteFees` *
+`requested CUs set by SetComputeUnitLimit`.
+
+If the payer does not have enough balance, the transaction is not scheduled and
+fails with an error `InsufficientFunds`. Consider this case the same as if the
+payer does not have enough funds to pay `other fees`.
+
+Before processing the message, we check if any loaded account has associated
+PRAW fees. For all accounts with PRAW fees, the fees paid should be greater than
+the required fees. In case of overpayment, the difference is stored in a
+variable and will be paid back to the payer in any case. In case the PRAW fees
+are insufficiently paid or not paid then the transaction fails with an error
+`ProgramRebatableAccountWriteFeesNotPaid`. The payer will pay just `other fees`.
+This failure is before the execution of the transaction.
+
+The PRAW fees minus rebates and minus overpayment are transferred to the
+respective accounts lamport balance, from where the owner can collect them and
+then the rest would be transferred back to the payer.
+
+### Syscall to change PRAW fees
+
+PRAW fees should be considered as set and forget kind of fees by dapp
+developers. They are not aimed to control congestion over the network instead
+they aim to reduce spamming.
+
+A new syscall will be added to update the PRAW fees:
+
+```
+fn update_program_rebatable_account_write_fees(
+  account: Pubkey, 
+  microlamports_per_requested_cu: u64
+) -> bool;
+```
+
+The syscall will take as arguments the account on which fee should be updated
+and the microlamports per cu that should be charged and return a true if the
+changes were successful else will return false.
+
+The accounts database will need to store the value of microlamports per
+requested CU for the account. Changes to update the PRAW fees could be done only
+by the authority program. All programs will need to implement special
+instructions so that the program admin can use this syscall to update PRAW fees
+on the required accounts. Program admin should always be signer for these
+instructions.
```

Comment:
nit: not always, so I suggest removing for clarity. for example this could be done on account initialization when no signer is required
```suggestion
on the required accounts.
```